<a href="https://colab.research.google.com/github/Jeronimo-Giraldo/Analitica-de-negocios/blob/main/Reto_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0.Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd

#Establecimento y evalución del modelo
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

#Carga de los archivos desde el drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1.Se procede con la carga de datos

In [ ]:
nxl="/content/2. TechSalesReport_Int_L.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0)
XDB['Business'] = XDB['Business'].replace({'Hardware': 0, 'Software': 1})
XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})
XDB = XDB.drop(['Sales_Rep', 'Personality'], axis=1)
print(XDB)

       Business  Age  Female  Years  College  Certificates  Feedback  Salary  \
0             0   59       1      2        1             1      2.01   70200   
1             0   52       0     10        1             4      3.64  133000   
2             1   47       1      1        1             1      3.88   52600   
3             0   61       0      2        1             3      2.70   96000   
4             1   39       0      1        0             5      3.44  122000   
...         ...  ...     ...    ...      ...           ...       ...     ...   
21985         1   44       0      1        1             4      1.80  128000   
21986         1   23       1      6        1             6      1.77   53400   
21987         0   48       1      4        1             0      2.46   50400   
21988         1   29       0      4        1             2      3.68   70000   
21989         1   23       1      2        1             1      2.13   47400   

       NPS  
0        5  
1       10  


<ipython-input-4-e4595f05cf84>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['Business'] = XDB['Business'].replace({'Hardware': 0, 'Software': 1})
<ipython-input-4-e4595f05cf84>:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})


2.Seleccionamos las variables

In [ ]:
# Definir XD e yd
XD=XDB[['Business', 'Age', 'Years', 'Female', 'Feedback', 'Salary', 'Certificates','NPS']]
yd=XDB['College']
print(XD)
print(yd)

       Business  Age  Years  Female  Feedback  Salary  Certificates  NPS
0             0   59      2       1      2.01   70200             1    5
1             0   52     10       0      3.64  133000             4   10
2             1   47      1       1      3.88   52600             1    8
3             0   61      2       0      2.70   96000             3    6
4             1   39      1       0      3.44  122000             5    7
...         ...  ...    ...     ...       ...     ...           ...  ...
21985         1   44      1       0      1.80  128000             4    7
21986         1   23      6       1      1.77   53400             6    8
21987         0   48      4       1      2.46   50400             0    4
21988         1   29      4       0      3.68   70000             2   10
21989         1   23      2       1      2.13   47400             1    5

[21990 rows x 8 columns]
0        1
1        1
2        1
3        1
4        0
        ..
21985    1
21986    1
21987    1

3.Se procede con la implementacion del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)

#Despliegue del arbol
from sklearn.tree import export_graphviz #Exportar
from pydotplus import graph_from_dot_data #Crear grafico

vs=['Business', 'Age', 'Years', 'Female', 'Feedback', 'Salary', 'Certificates','NPS']

dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('1. ArbolPreApr.png')

True

4.Desempeño del modelo

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)


VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Exactitud=(VP+VN)/(VP+VN+FP+FN) #Funcionamiento general
print('Exactitud',Exactitud)
Sensibilidad=VP/(VP+FN) #Como se comporto el modelo con los aprobados
print('Sensibilidad',Sensibilidad)
Especificidad=VN/(VN+FP) #Como se comporto el modelo con los negados
print('Especificidad',Especificidad)
Precision=VP/(VP+FP) #acierto clasificando los positivos
print('Precision',Precision)
TasaError=(FP+FN)/(VP+VN+FP+FN) #Equivocación general
print('TasaError',TasaError)
PredNeg=VN/(VN+FN) #Como funciona el modelo negando creditos
print('PredNeg',PredNeg)

[[  114  4356]
 [   69 17451]]
Exactitud 0.7987721691678036
Sensibilidad 0.9960616438356165
Especificidad 0.025503355704697986
Precision 0.8002476269087908
TasaError 0.20122783083219645
PredNeg 0.6229508196721312


5.Número de personas y de nodos puros para el arbol

In [ ]:
# Total de personas
total_personas = len(XDB)

# Número de personas con título universitario (inscritas)
num_inscritas = XDB['College'].sum()

# Número de personas sin título universitario (no inscritas)
num_no_inscritas = total_personas - num_inscritas

# Imprimir los resultados
print(f"Total de personas en la base de datos: {total_personas}")
print(f"Número de personas con título universitario (inscritas): {num_inscritas}")
print(f"Número de personas sin título universitario (no inscritas): {num_no_inscritas}")

Total de personas en la base de datos: 21990
Número de personas con título universitario (inscritas): 17520
Número de personas sin título universitario (no inscritas): 4470


6.Intervalo titulo universitario

In [ ]:
filtered_data = XDB[(XDB['Salary'] <= 64100) & (XDB['Salary'] > 47100) & (XDB['Age'] > 39.5) & (XDB['Certificates'] > 1.5)]
num_college_degree = filtered_data['College'].sum()
print(f"Número de personas con título universitario en el intervalo: {num_college_degree}")
filtered_data = XDB[(XDB['Salary'] <= 64100) & (XDB['Salary'] > 47100) & (XDB['Age'] > 39.5) & (XDB['Certificates'] <= 1.5)]
num_no_college_degree = len(filtered_data) - filtered_data['College'].sum()
print(f"Número de personas sin título universitario en el intervalo: {num_no_college_degree}")

Número de personas con título universitario en el intervalo: 986
Número de personas sin título universitario en el intervalo: 356


7.pronostico en un archivo de excel

In [ ]:
import pandas as pd
# Instala la librería openpyxl si no la tienes:
# Hacer un DataFrame con los pronósticos
# Suponiendo que 'ydp' contiene los pronósticos del modelo
#XD contiene los datos de entrada
df_pronostico = pd.DataFrame({'Pronostico': ydp}, index=XD.index)

# Unir el DataFrame de pronósticos con los datos originales (opcional)
df_completo = pd.concat([XD, df_pronostico], axis=1)

# Guardar el DataFrame en un archivo de Excel
df_completo.to_excel('informe_pronostico.xlsx', sheet_name='Pronosticos', index=False)

8.Analisis del ejercicio

Se implementó un modelo de Árbol de Decisión utilizando ocho variables de entrada: Business, Age, Years, Female, Feedback, Salary, Certificates y NPS. La variable de salida fue College, que evalúa si una persona posee un título universitario. La base de datos analizada incluyó un total de 21,990 registros, donde 4,470 personas no tenían título universitario y 17,520 sí lo tenían. El modelo alcanzó una exactitud de casi 80%, superando el umbral aceptable del 75%, lo que refleja un desempeño sólido en la clasificación de los datos,en los indicadores de rendimiento el modelo presentó una sensibilidad de casi 100%, demostrando una alta capacidad para identificar correctamente a las personas con título universitario, Sin embargo, la especificidad fue de apenas 3%, lo que indica una baja efectividad al clasificar los casos negativos. La precisión positiva alcanzó un valor cercano al 80%, mientras que la precisión negativa fue de aproximadamente 62%, evidenciando una tendencia del modelo a clasificar mayormente casos positivos. Posteriormente, se utilizó el modelo para predecir la preaprobación universitaria en los 4,470 registros de personas sin títul dónde la mayoría fueron clasificados erróneamente como poseedores de un título universitario. Además, se observó que ningún nodo del árbol es puro, lo que significa que en todos los nodos hoja existe una mezcla de ambas categorías con y sin título universitario, lo que limita la precisión del modelo.